### DISCUSSION TITLE
Makefile conditionals?

### DISCUSSION THREAD

**Discussion by SirWumpus** [ioccc-src/mkiocccentry#862] [General]
What is the policy concerning Makefile conditionals?  In particular for handling target platform building eg. Cygwin, *BSD, Linux.  GNU Make and BSD Make (as used for [PkgSrc](http://pkgsrc.org/)) have slightly different conditional syntax.  Consider the BSD Make fragment:

```
.POSIX :

.if ${.MAKE.OS} == "Cygwin"
O := .obj
E := .exe
.else
O := .o
E :=
.endif

.SUFFIXES :
.SUFFIXES : .c .h .i $O $E
```

* Will the IOCCC Makefiles handles these target platform issues?  
* Will there be minor allowances for contestant Makefile (if documented of course)?
* Will multiple Makefiles need to be maintained by platform?

**Comment by xexyl** [comment]
> What is the policy concerning Makefile conditionals?  In particular for handling target platform building eg. Cygwin, *BSD, Linux.  GNU Make and BSD Make (as used for [PkgSrc](http://pkgsrc.org/)) have slightly different conditional syntax.  Consider the BSD Make fragment:
> 
> 
> 
> ```
> 
> .POSIX :
> 
> 
> 
> .if ${.MAKE.OS} == "Cygwin"
> 
> O := .obj
> 
> E := .exe
> 
> .else
> 
> O := .o
> 
> E :=
> 
> .endif
> 
> 
> 
> .SUFFIXES :
> 
> .SUFFIXES : .c .h .i $O $E
> 
> ```
> 
> 
> 
> * Will the IOCCC Makefiles handles these target platform issues?  
> 
> * Will there be minor allowances for contestant Makefile (if documented of course)?
> 
> * Will multiple Makefiles need to be maintained by platform?
> 
> 

The answer to the second question is in the example Makefile: you're allowed to modify it as you see fit though the tool does try and find specific rules that are noted as existing or not in the json file. 

I have no idea about the other questions except to say that the winning entry Makefiles do have conditionals for different compilers so I presume that if necessary they can be updated to have other conditionals (syntax differences notwithstanding?). 


Interesting thoughts.

**Comment by lcn2** [comment]
Hello @SirWumpus,

As you created this discussion in the mkiocccentry repo, we will reply in the mkiocccentry repo toolset context.

The mkiocccentry repo assumes a Gnu Make compatible make tool.  We do not plan to support the building of this repo using the "traditional" Bell Labs make, nor a "traditional" BSD-style make, nor bmake, nor System V make, etc.

Regarding Cygwin: We are told that some Cygwin installations have a Gnu Make compatible make tool.   We do not plan to test compiling and running the mkiocccentry repo toolset under Cygwin, so we do not know if it would work or not.  We do not plan to support mkiocccentry under Cygwin.

While IOCCC submissions may come from a wide variety of backgrounds, the building of the compressed tarball needs to be done on a system that is able to compile the mkiocccentry repo and run the `mkiocccentry(1)` tool. 

Please keep in mind that someone who is build and testing their IOCCC submission under any of a wide variety if platforms, will need to copy their submission onto a system where the the mkiocccentry repo and been compiled and where the `mkiocccentry(1)` can run. 

## UPDATE 0

P.S., Your question, @SirWumpus, might imply you are (also?) asking about the next IOCCC and/or asking about the new system of compiling IOCCC entries.

If that is case, @SirWumpus, please open a discussion [temp-test-ioccc repo discussion](https://github.com/ioccc-src/temp-test-ioccc/discussions).

**Reply by SirWumpus** [reply]
> ... nor a "traditional" BSD-style make, nor bmake, nor System V make, etc.

I tend to work with BSD primarily (and yes I do install `gmake` to build stuff for those that think GNU is a cat's meow).

I also try to test my builds on Cygwin (because its often handy and a nice plus) and it does offer a `bmake` (which I never needed before this question).  I also more reluctantly try to test on some Linux (Alpine or Rocky) for max. portability testing.

As for IOCCC supporting Cygwin, its not a big deal.  Its more likely someone will try Windows Linux subsystem (insert a phellem emoji). 

> ... imply you are (also?) asking about the next IOCCC and/or asking about the new system of compiling IOCCC entries.

Not really so much.  I'm sure what ever tool chain is put forward will be sufficiently obfuscated in manner befitting the contest.

I patiently await() then next contest with baited garlic breath.

**Reply by xexyl** [reply]
As I noted in the other issue if you think an FAQ entry about this is useful, Landon, I would be happy to do it. 

It seems like it might be a good idea?

**Reply by lcn2** [reply]
Yes, @xexyl please.

**Reply by xexyl** [reply]
> Yes, @xexyl please.

Okay. Perhaps you would note it in the other thread? 

(I will get to this in the next week or so. I am still not sure how much I can get done before my surgery as a lot of things are going on. 

I will certainly need some rest but it shouldn't take much time away from the contest. 

Of course it's possible I will get to this task before. I would like to manage it tomorrow morning but it depends on a number of factors. 

Thanks and please remind me on this if I don't get to it!)

**Comment by xexyl** [comment]
About to open a pull request for this .. just have to generate the html file, upload to server (scp) and then check the generated output: links etc. If all good I will commit and open a pull request. That's all I'll do here today I'm afraid but that's better than the past some days. I do hope that I can do more soon!

## UPDATE 0

.... well that's just great. Saw for first time something strange in scp: 'stalled'. Will try rsync ..

**Comment by xexyl** [comment]
> About to open a pull request for this .. just have to generate the html file, upload to server (scp) and then check the generated output: links etc. If all good I will commit and open a pull request. That's all I'll do here today I'm afraid but that's better than the past some days. I do hope that I can do more soon!
> 
> ## UPDATE 0
> .... well that's just great. Saw for first time something strange in scp: 'stalled'. Will try rsync ..

See https://github.com/ioccc-src/temp-test-ioccc/pull/2286/commits/1023fe227fc31bd3d51cb3aff0295d2e236b408c.
